ports/rp2/machine_uart.c

```diff
@@ -179,31 +212,52 @@ STATIC const mp_rom_map_elem_t machine_uart_locals_dict_table[] = {
     { MP_ROM_QSTR(MP_QSTR_write), MP_ROM_PTR(&mp_stream_write_obj) },
 
     { MP_ROM_QSTR(MP_QSTR_sendbreak), MP_ROM_PTR(&machine_uart_sendbreak_obj) },
+
+    { MP_ROM_QSTR(MP_QSTR_INV_TX), MP_ROM_INT(UART_INVERT_TX) },
+    { MP_ROM_QSTR(MP_QSTR_INV_RX), MP_ROM_INT(UART_INVERT_RX) },
+
 };
 STATIC MP_DEFINE_CONST_DICT(machine_uart_locals_dict, machine_uart_locals_dict_table);
 
 STATIC mp_uint_t machine_uart_read(mp_obj_t self_in, void *buf_in, mp_uint_t size, int *errcode) {
     machine_uart_obj_t *self = MP_OBJ_TO_PTR(self_in);
-    // TODO support timeout
+    uint64_t t = time_us_64() + (uint64_t)self->timeout * 1000;
     uint8_t *dest = buf_in;
-    for (size_t i = 0; i < size; ++i) {
+
+    for (size_t i = 0; i < size; i++) {
+        // Wait for the first/next character
         while (!uart_is_readable(self->uart)) {
+            if (time_us_64() > t) {  // timed out
+                return i;  // return silently
+            }
             MICROPY_EVENT_POLL_HOOK
         }
         *dest++ = uart_get_hw(self->uart)->dr;
+        t = time_us_64() + (uint64_t)self->timeout_char * 1000;
```

Comment:
Might be worth caching `uint64_t timeout_char_us = (uint64_t)self->timeout_char * 1000` outside this loop.

Reply:
Done